### DISCUSSION TITLE
Unable to Find Entry Point in DLL When Using DllImport in Avalonia UI

### DISCUSSION THREAD

**Discussion by Samdani1** [AvaloniaUI/Avalonia#12477] [Q&A]
I'm currently working on an Avalonia UI project where I need to call a function from a custom DLL using the DllImport attribute. However, I'm encountering an issue where I'm getting the error message:

System.EntryPointNotFoundException: 'Unable to find an entry point named 'GiveMeNumber' in DLL 'LibraryAdd'.'

Here's the relevant code snippet from my Avalonia UI project:
```
[DllImport(@"LibraryAdd", EntryPoint = "GiveMeNumber", SetLastError = true, CharSet = CharSet.Unicode)]
public static extern int GiveMeNumber();

private void PointerPressed(object sender, PointerPressedEventArgs e)
{
GiveMeNumber();
}
```
I have verified that the DLL 'LibraryAdd.dll' exists in the same directory as my executable. The function 'GiveMeNumber' is supposed to be exported from the DLL, and I successfully accessed it in a Console Application with the same codebase. However, when I try to call the same function from my Avalonia UI application, I encounter the error.

What could be causing this issue? How can I resolve it and successfully call the GiveMeNumber function from my Avalonia UI application?

I would greatly appreciate any insights or solutions to this problem. Thank you in advance for your help!

**Comment by kekekeks** [comment]
Check if the function is resolvable using NativeLibrary APIs if you use the full path.

Check if the function name actually does exist in the dll being loaded using [depends.exe](https://www.dependencywalker.com/) (windows) or `objdump -T <libname> | grep Base` (Linux).

**Reply by Samdani1** [reply]
Yes! The name does exist in the dll and I am able access it in a Console Application as well with the same codebase.

![image](https://github.com/AvaloniaUI/Avalonia/assets/85519060/75736676-c96d-4367-8dd7-ca909924ee1b)

**Reply by kekekeks** [reply]
Check if there are no NativeLibrary.SetDllImportResolver shenanigans. You can verify that by trying to call that API for your assembly just before P/Invoke, it will fail if it was called before.

**Reply by Samdani1** [reply]
I have attempted to call the GiveMeNumber function just before the P/Invoke inside Program.cs, but the error still persists. It seems that my approach is not resolving the issue related to SetDllImportResolver.

Here's the relevant part of my code:
```
internal class Program
{
    [DllImport(@"LibraryAdd", EntryPoint = "GiveMeNumber", SetLastError = true, CharSet = CharSet.Unicode)]
    public static extern int GiveMeNumber();

    // Verification code for SetDllImportResolver
    static Program()
    {
        // Attempt to call the API just before P/Invoke
        int result = GiveMeNumber();
        Console.WriteLine($"API Call Result: {result}");
    }

    // Initialization code. Don't use any Avalonia, third-party APIs or any
    // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
    // yet and stuff might break.
    [STAThread]
    public static void Main(string[] args)
    {
        BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);
    }

    // Avalonia configuration, don't remove; also used by the visual designer.
    public static AppBuilder BuildAvaloniaApp()
        => AppBuilder.Configure<App>()
            .UsePlatformDetect()
            .WithInterFont()
            .LogToTrace();
}
```
The verification code in the static constructor should provide some insight into whether SetDllImportResolver is functioning as expected. However, despite this attempt, the error related to EntryPointNotFoundException for the GiveMeNumber function persists.

Am I following the right path to diagnose and resolve the issue? Is there anything else I should consider to successfully call GiveMeNumber from my Avalonia application?

Thank you for your assistance!

**Comment by maxkatz6** [comment]
Duplicate https://github.com/AvaloniaUI/Avalonia/issues/10046
